Compare TUV-X Tridiagonal Solver Code with LAPACKE's implementation (#97)
NCAR/tuv-x: CMakeLists.txt

> # ##############################################################################

Reviewer: ```suggestion
################################################################################
```

# Preamble

cmake_minimum_required(VERSION 3.17)

project(
  tuv-x
  VERSION 0.8.0
  LANGUAGES Fortran CXX C)

set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
set(CMAKE_MODULE_PATH "${CMAKE_MODULE_PATH};${PROJECT_SOURCE_DIR}/cmake")

set(CMAKE_USER_MAKE_RULES_OVERRIDE
    "${PROJECT_SOURCE_DIR}/cmake/set_defaults.cmake")

if(NOT CMAKE_BUILD_TYPE)

>   set(CMAKE_BUILD_TYPE
>       "Release"
>       CACHE
>         STRING
>         "Choose the type of build, options are: Debug Release RelWithDebInfo MinSizeRel."
>         FORCE)

Reviewer 2: Not really sure how to best handle the formatting but I think this would be a nice middle ground and uses the built-in list of build types (https://cmake.org/cmake/help/latest/variable/CMAKE_CONFIGURATION_TYPES.html).
```suggestion
  set(CMAKE_BUILD_TYPE "Release"
      CACHE STRING
        "Choose the type of build, options are: ${CMAKE_CONFIGURATION_TYPES}."
      FORCE)
```

endif(NOT CMAKE_BUILD_TYPE)

> message(
>   STATUS
>     "CMake build configuration for tuv-x(${CMAKE_BUILD_TYPE}) ${PROJECT_VERSION}"
> )

Reviewer 2: Not sure why this was reformatted but is there a reason the original was not ideal?  All for getting fid of the spaces between `message`, `(`, and `STATUS` though.
```suggestion
message(STATUS "CMake build configuration for tuv-x(${CMAKE_BUILD_TYPE}) ${PROJECT_VERSION}")
```

Author: Seems like my code editor reformats files whenever I save them. I was just debugging the `FindLAPACKE` by printing out the include paths found by the function and it reformats the file whenever I delete those print statements and do `:wq`. I will try to turn this off for cmake files.

if(${CMAKE_VERSION} VERSION_LESS "3.21")
  if(CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
    set(PROJECT_IS_TOP_LEVEL TRUE)
  else()
    set(PROJECT_IS_TOP_LEVEL FALSE)
  endif()
endif()

> # ##############################################################################

Reviewer: ```suggestion
################################################################################
```

project(
  tuv-x

>   VERSION 0.8.0

Reviewer 3: Do we want to downgrade the version here?

Author: I did not make this change, so I am not sure. Maybe this is a question for @K20shores ?

Reviewer: Well that's definitely me doing that by accident. We can undo that

include(CMakeDependentOption)
option(TUVX_ENABLE_MPI "Enable MPI parallel support" OFF)
cmake_dependent_option(TUVX_ENABLE_OPENMP "Enable OpenMP support" OFF
                       "TUVX_ENABLE_MPI" OFF)
option(TUVX_ENABLE_TESTS "Build tests" ON)

> option(TUVX_ENABLE_BENCHMARK "Build benchmark examples" ON)

Reviewer 4: Do we mean to set this option `ON`as default?

Author: I'm not sure. Maybe this is a question for @sjsprecious @mattldawson @K20shores ?

Reviewer 5: I think this should probably be off by default. Users will probably not be interested in running the benchmark tests

Author: @mattldawson  This seems to be off by default (even though it set to be on in the cmake file). I had to pass in `- D TUVX_ENABLE_BENCHMARK=ON` for cmake to generate the build files for the benchmarking code.

Author: @boulderdaze I turned it off in my latest commit.

endif()

# copy data
if(TUVX_ENABLE_TESTS)
  add_custom_target(
    copy-data ALL
    COMMAND ${CMAKE_COMMAND} -E copy_directory ${CMAKE_CURRENT_SOURCE_DIR}/data
            ${CMAKE_BINARY_DIR}/data)
  add_custom_target(
    copy-tool ALL
    COMMAND ${CMAKE_COMMAND} -E copy_directory ${CMAKE_CURRENT_SOURCE_DIR}/tool
            ${CMAKE_BINARY_DIR}/tool)
  add_custom_target(
    copy-examples ALL
    COMMAND ${CMAKE_COMMAND} -E copy_directory
            ${CMAKE_CURRENT_SOURCE_DIR}/examples ${CMAKE_BINARY_DIR}/examples)
endif()

> # ##############################################################################

Reviewer: ```suggestion
################################################################################
```

# Dependencies

include(cmake/dependencies.cmake)

> # ##############################################################################

Reviewer: ```suggestion
################################################################################
```

add_executable(tuv-x src/tuvx.F90 version.F90)

set_target_properties(tuv-x PROPERTIES LINKER_LANGUAGE Fortran)

if(NOT BUILD_SHARED_LIBS)
  set_target_properties(tuv-x PROPERTIES POSITION_INDEPENDENT_CODE ON)
endif()

> target_link_libraries(tuv-x PUBLIC musica::tuvx yaml-cpp::yaml-cpp
>                                    ${BLAS_LIBRARIES} ${LAPACK_LIBRARIES})

Reviewer: ```suggestion
target_link_libraries(tuv-x PUBLIC musica::tuvx yaml-cpp::yaml-cpp)
```

If you would, please publicly link `${BLAS_LIBRARIES} ${LAPACK_LIBRARIES}` to musica::tuvx. We wouldn't need to add links to `${BLAS_LIBRARIES} ${LAPACK_LIBRARIES}` anywhere else in the cmake files if we do that

target_include_directories(
  tuv-x PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include>
               $<INSTALL_INTERFACE:${TUVX_INSTALL_INCLUDE_DIR}>)

if(TUVX_ENABLE_OPENMP)
  target_link_libraries(tuv-x PUBLIC OpenMP::OpenMP_Fortran)
endif()

> # ##############################################################################

Reviewer: ```suggestion
################################################################################
```

# TUV-x docs

if(TUVX_BUILD_DOCS)
  add_subdirectory(docs)
endif()

> # ##############################################################################

Reviewer: ```suggestion
################################################################################
```

    include(CodeCoverage)
    if(${CMAKE_Fortran_COMPILER_ID} MATCHES "GNU")
      setup_target_for_coverage_lcov(
        NAME
        coverage
        EXECUTABLE
        "ctest"
        EXCLUDE
        "${PROJECT_SOURCE_DIR}/test/*"
        BASE_DIRECTORY

>         "${PROJECT_SOURCE_DIR}/src")

Reviewer: Please revert this

    endif()
  endif()

  enable_testing()
  add_subdirectory(test)
endif()

> # ##############################################################################

Reviewer: ```suggestion
################################################################################
```

# Packaging

# only include packaging if we are the top level project being built
if(PROJECT_IS_TOP_LEVEL)
  add_subdirectory(packaging)
endif()

> # ##############################################################################

Reviewer: ```suggestion
################################################################################
```
